### PR TITLE
Fix substitution of type params

### PR DESCRIPTION
**Pull request by Julow**
Fix https://github.com/ocaml/odoc/issues/538 and https://github.com/ocaml/odoc/issues/462

This substitutes recursively type variables when encountering a "replaced" type.

### CODE REVIEW

File: test/xref2/gen.ml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+let generate input =
+  run "odoc man-generate -o %S %s" out_dir (Filename.quote input)
+
+let () =
+  for i = 1 to Array.length Sys.argv - 1 do
+    generate Sys.argv.(i)
+  done;
+  list_dir_rec [] out_dir
+  |> List.iter (fun p -> run "COLUMNS=80 man -l %s | col -b" p)
```
Comment (Julow) [inline test/xref2/gen.ml:31]: These tests may be very fragile because they depend on `man` and `col`, which are very complicated programs. I'm replacing that with Mdx tests instead.

File: src/xref2/subst.ml [hunk 2]
```diff
@@ -571,11 +636,13 @@ and extension_constructor s c =
...
-  {
-    e with
-    type_path = type_path s e.type_path;
-    constructors = List.map (extension_constructor s) e.constructors;
-  }
+  let type_path =
+    match type_path s e.type_path with
+    | Not_replaced p -> p
+    | Replaced (TypeExpr.Constr (p, _), _) -> p
+    | Replaced _ -> (* What else is possible ? *) assert false
```
Comment (Julow) [inline src/xref2/subst.ml:643]: I'm not sure what we should do here. More tests are required

File: test/xref2/subst/test.md [hunk 3]
```diff
@@ -76,10 +76,141 @@ type vv/5 = local(SubstituteMe/2,false).v
...
+  |}
+- : Component.Signature.t =
+module type S/69 = sig
+  type t/70
+  val map/71 : ([a] t/70) -> ((a) -> b) -> [b] t/70
+   (removed=[])end
+module M/68 : sig
+  type t/72
+  include : S/69 with [root(S/69).t = [(alias (poly_var [ `A of (a * b) ]) b)] t/72] (sig =
+    val map/73 : ([(alias (poly_var [ `A of (a * b) ]) b)] t/72) -> ((a) -> b) -> [(alias (poly_var [ `A of (b * b) ]) b)] t/72
```
Comment (Julow) [inline test/xref2/subst/test.md:212]: I've found an edge case, the last type is wrong: `(alias (poly_var [ `A of (b * b) ]) b)`.
We should either rename `b` to `c` in this type or choose a free variable before substituting.
What do you think ?
